src/jxl.imageio/jxlinput.cpp

```diff
@@ -0,0 +1,358 @@
+
+#include <algorithm>
+#include <cassert>
+#include <cstdio>
+
+#include <OpenImageIO/filesystem.h>
+#include <OpenImageIO/fmath.h>
+#include <OpenImageIO/imageio.h>
+#include <OpenImageIO/tiffutils.h>
+
+#include <jxl/decode.h>
+#include <jxl/decode_cxx.h>
+#include <jxl/resizable_parallel_runner_cxx.h>
+
+OIIO_PLUGIN_NAMESPACE_BEGIN
+
+#define DBG if (0)
+
+class JxlInput final : public ImageInput {
+public:
+    JxlInput() { init(); }
+    ~JxlInput() override { close(); }
+    const char* format_name(void) const override { return "jxl"; }
+    int supports(string_view feature) const override
+    {
+        return (feature == "exif" || feature == "ioproxy");
+    }
+    bool valid_file(Filesystem::IOProxy* ioproxy) const override;
+
+    bool open(const std::string& name, ImageSpec& spec) override;
+    bool open(const std::string& name, ImageSpec& spec,
+              const ImageSpec& config) override;
+    bool read_native_scanline(int subimage, int miplevel, int y, int z,
+                              void* data) override;
+    bool close() override;
+
+    const std::string& filename() const { return m_filename; }
+
+private:
+    std::string m_filename;
+    int m_next_scanline;  // Which scanline is the next to read?
+    uint32_t m_channels;
+    JxlDecoderPtr m_decoder;
+    JxlResizableParallelRunnerPtr m_runner;
+    std::unique_ptr<ImageSpec> m_config;  // Saved copy of configuration spec
+    std::vector<uint8_t> m_icc_profile;
+    std::vector<float> m_pixels;
+    // std::vector<uint8_t> m_pixels;
+
+    void init()
+    {
+        ioproxy_clear();
+        m_config.reset();
+        m_decoder = nullptr;
+        m_runner  = nullptr;
+    }
+
+    void close_file() { init(); }
+};
+
+// Export version number and create function symbols
+OIIO_PLUGIN_EXPORTS_BEGIN
+
+OIIO_EXPORT int jxl_imageio_version = OIIO_PLUGIN_VERSION;
+
+OIIO_EXPORT const char*
+jxl_imageio_library_version()
+{
+    return "libjxl " OIIO_STRINGIZE(JPEGXL_MAJOR_VERSION) "." OIIO_STRINGIZE(
+        JPEGXL_MINOR_VERSION) "." OIIO_STRINGIZE(JPEGXL_PATCH_VERSION);
+}
+
+OIIO_EXPORT ImageInput*
+jxl_input_imageio_create()
+{
+    return new JxlInput;
+}
+
+OIIO_EXPORT const char* jxl_input_extensions[] = { "jxl", nullptr };
+
+OIIO_PLUGIN_EXPORTS_END
+
+bool
+JxlInput::valid_file(Filesystem::IOProxy* ioproxy) const
+{
+    DBG std::cout << "JxlInput::valid_file()\n";
+
+    // Check magic number to assure this is a JPEG file
+    if (!ioproxy || ioproxy->mode() != Filesystem::IOProxy::Read)
+        return false;
+
+    uint8_t magic[128] {};
+    const size_t numRead = ioproxy->pread(magic, sizeof(magic), 0);
+    if (numRead != sizeof(magic))
+        return false;
+
+    JxlSignature signature = JxlSignatureCheck(magic, sizeof(magic));
+    switch (signature) {
+    case JXL_SIG_CODESTREAM:
+    case JXL_SIG_CONTAINER: break;
+    default: return false;
+    }
+
+    DBG std::cout << "JxlInput::valid_file() return true\n";
+    return true;
+}
+
```

Comment:
I know it's picky, but would you mind using our standard of 3 blank lines between top-level functions? It's not critical, but I think it makes the code easier to visually parse.

Reply:
Corrected in the latest iteration.